Change UInt64Index._na_value from 0 to np.nan

jschendel opened the pull request:
Prerequisite for #18300

- [X] closes #18398
- [X] closes #18400
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Summary:
- Changed `UInt64Index._na_value` from 0 to `np.nan`

- Added a `dtype` parameter to `_try_convert_to_int_index` to skip the initial attempt to coerce to `Int64Index` in cases where we really want `UInt64Index` (fix for #18400).

- Moved `test_where` and `test_where_array_like` from `TestInt64Index` to the `NumericInt` base class for more generic coverage, and forced it to check that things get coerced to `Float64Index`.  The way it was originally written raised a ValueError due to the 0 -> `np.nan` change.

jschendel committed:
Change UInt64Index._na_value from 0 to np.nan

jschendel commented on pandas/tests/indexes/test_base.py:
```diff
@@ -201,6 +201,25 @@ def __array__(self, dtype=None):
+
+        # fall back to Float64Index
+        data = [0.0, 1.1, 2.2, 3.3]
+        expected = Float64Index(data)
```
In light of Case III in #15832, do we actually want this behavior, or should it really coerce to `Int64Index([0, 1, 2, 3])`?  Originally wrote this test case to make sure all relevant paths were hit, without seeing #15832.

jreback replied:
these should raise as they are passing a dtype that is non convertible

jschendel replied:
Removed the `Float64Index` portion of the test.  Was able to get it to raise, but ended up breaking other things.  Not immediately sure of the fix, but seems outside the scope of this PR, though can look into it more if need be. None of the code I wrote actually depends on that behavior; originally included it to make sure all code paths were tested.

jschendel committed:
update test

jschendel commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -108,7 +108,7 @@ Bug Fixes
 ^^^^^^^^^^
 
--
+- Bug in :class:`Index` constructor with `dtype='uint64'` where int-like floats were not coerced to :class:`UInt64Index` (:issue:`18400`)
```
Added a separate entry for 18398 under "Backwards incompatible API changes"

jschendel commented on pandas/core/indexes/base.py:
```diff
@@ -685,12 +685,13 @@ def _try_convert_to_int_index(cls, data, copy, name):
-                return Int64Index(res, copy=copy, name=name)
-        except (OverflowError, TypeError, ValueError):
-            pass
+        if not is_unsigned_integer_dtype(dtype):
```
done

jschendel commented:
Regarding the `test_where` and `test_where_arraylike` tests:  having both seemed redundant, as there was a bit of overlap, so I combined them into a single test.  Some notes:
- replaced `notna` with a list of `True`, since that's all that `notna` returned, so calling `notna` seemed unnecessarily convoluted.
- Moved the tests from `NumericInt` to `Numeric` since the new version of the test also passed for `RangeIndex`.  Deleted the `RangeIndex` specific version of the tests.
- Went through and made similar changes to other `test_where` and `test_where_arraylike` tests that were essentially doing the same thing.

jschendel committed:
review edits

jreback commented:
can you rebase. ready to go?

jorisvandenbossche commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -36,7 +36,7 @@ Backwards incompatible API changes
 
 - :func:`Series.fillna` now raises a ``TypeError`` instead of a ``ValueError`` when passed a list, tuple or DataFrame as a ``value`` (:issue:`18293`)
--
+- The default NA value for :class:`UInt64Index` has changed from 0 to ``NaN`` (:issue:`18398`)
```
What user visible change has this? 
Now it sounds a bit vague

jschendel replied:
It impacts things that  mask with NA under the hood, such as `.where` 

Previous behavior:
```python
In [3]: idx = pd.UInt64Index(range(5))

In [4]: idx
Out[4]: UInt64Index([0, 1, 2, 3, 4], dtype='uint64')

In [5]: idx.where(idx > 3)
Out[5]: UInt64Index([0, 0, 0, 0, 4], dtype='uint64')
```

New behavior:
```python
In [3]: idx = pd.UInt64Index(range(5))

In [4]: idx.where(idx > 3)
Out[4]: Float64Index([nan, nan, nan, nan, 4.0], dtype='float64')
```

Updating the whatsnew to make this more clear.

jschendel committed:
clarified whatsnew

jschendel commented:
@jreback : this should be ready to merge once green, assuming that the whatsnew change I made to address the comment by @jorisvandenbossche is acceptable.

jreback commented on pandas/tests/indexes/test_category.py:
```diff
@@ -269,28 +269,19 @@ def f(x):
 
-    def test_where(self):
+    @pytest.mark.parametrize('klass', [list, tuple, np.array, pd.Series])
+    def test_where(self, klass):
```
we should prob move all of the test_where test to test_base and use the indices fixture to avoid the code repetition (new issue though)

jreback commented:
thanks @jschendel very nice!
